nav2_collision_monitor/src/pointcloud.cpp

```diff
@@ -0,0 +1,126 @@
+  min_height_(0.0), max_height_(0.0), data_(nullptr)
+{
+  RCLCPP_INFO(logger_, "[%s]: Creating PointCloud", source_name_.c_str());
+}
+
+PointCloud::~PointCloud() {
+  RCLCPP_INFO(logger_, "[%s]: Destroying PointCloud", source_name_.c_str());
+
+  tf_buffer_.reset();
+  data_sub_.reset();
+}
+
+void PointCloud::configure()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  std::string source_topic;
+
+  getParameters(source_topic);
+
+  rclcpp::QoS pointcloud_qos = rclcpp::SensorDataQoS();  // set to default
+  data_sub_ = node->create_subscription<sensor_msgs::msg::PointCloud2>(
+    source_topic, pointcloud_qos,
+    std::bind(&PointCloud::dataCallback, this, std::placeholders::_1));
+}
+
+void PointCloud::getData(const rclcpp::Time & curr_time, std::vector<Point> & data)
+{
+  // Ignore data from the source if it is not being published yet or
+  // not published for a long time
+  if (data_ == nullptr || !sourceValid(curr_time)) {
+    return;
+  }
+
+  // Obtaining the transform to get data from source_frame_id_ and time where it was obtained
+  // to base_frame_id_ frame and current time
+  tf2::Transform tf2_transform;
+  if (!getSourceBaseTransform(curr_time, data_stamp_, tf2_transform)) {
+    return;
+  }
+
+  sensor_msgs::PointCloud2ConstIterator<float> iter_x(*data_, "x");
+  sensor_msgs::PointCloud2ConstIterator<float> iter_y(*data_, "y");
+  sensor_msgs::PointCloud2ConstIterator<float> iter_z(*data_, "z");
+
+  // Refill data array with PointCloud points in base_frame_id_
+  for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z) {
+    // Transform point coordinates from source_frame_id_ -> to base_frame_id_
+    tf2::Vector3 p_v3_s(*iter_x, *iter_y, *iter_z);
+    tf2::Vector3 p_v3_b = tf2_transform * p_v3_s;
+
+    // Refill data array
+    if (p_v3_b.z() >= min_height_ && p_v3_b.z() <= max_height_) {
+      data.push_back({p_v3_b.x(), p_v3_b.y()});
+    }
+  }
+}
+
+void PointCloud::getParameters(std::string & source_topic)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  SourceBase::getParameters(source_topic);
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, source_name_ + ".min_height", rclcpp::ParameterValue(0.05));
+  min_height_ = node->get_parameter(source_name_ + ".min_height").as_double();
+  nav2_util::declare_parameter_if_not_declared(
+    node, source_name_ + ".max_height", rclcpp::ParameterValue(0.5));
+  max_height_ = node->get_parameter(source_name_ + ".max_height").as_double();
+}
+
+void PointCloud::dataCallback(sensor_msgs::msg::PointCloud2::ConstSharedPtr msg)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Obtaining source_frame_id_ from first message
+  if (source_frame_id_.length() == 0) {
+    source_frame_id_ = msg->header.frame_id;
+  }
+
+  data_ = msg;
+  data_stamp_ = node->now();
```

Comment:
This should be the message's stamp, after the initial `now()` at the start main function to run this behavior, no one else should be getting their own stamps. Use the message's header for the time it was captured, not when it was received

Reply:
Update for scan too

Reply:
That was the line remained after experiments, I've missed to roll back to `msg->header.stamp`. For scan, there is OK, only for pointcloud. Thank you for pointing this out!